### ISSUE TITLE
Using python generators and steps per epoch in Skorch

### ISSUE THREAD

**Issue by sushmit86** [skorch-dev/skorch#820]
I am trying to convert a keras code to Pytorch. The code is from the first edition of [Deep learning with Python](https://github.com/fchollet/deep-learning-with-python-notebooks/blob/master/first_edition/6.3-advanced-usage-of-recurrent-neural-networks.ipynb).

Here is the keras code

```
def generator(data, lookback, delay, min_index, max_index,
              shuffle=False, batch_size=128, step=6):
    if max_index is None:
        max_index = len(data) - delay - 1
    i = min_index + lookback
    while 1:
        if shuffle:
            rows = np.random.randint(
                min_index + lookback, max_index, size=batch_size)
        else:
            if i + batch_size >= max_index:
                i = min_index + lookback
            rows = np.arange(i, min(i + batch_size, max_index))
            i += len(rows)

        samples = np.zeros((len(rows),
                           lookback // step,
                           data.shape[-1]))
        targets = np.zeros((len(rows),))
        for j, row in enumerate(rows):
            indices = range(rows[j] - lookback, rows[j], step)
            samples[j] = data[indices]
            targets[j] = data[rows[j] + delay][1]
        yield samples, targets
lookback = 1440
step = 6
delay = 144
batch_size = 128

train_gen = generator(float_data,
                      lookback=lookback,
                      delay=delay,
                      min_index=0,
                      max_index=200000,
                      shuffle=True,
                      step=step, 
                      batch_size=batch_size)
val_gen = generator(float_data,
                    lookback=lookback,
                    delay=delay,
                    min_index=200001,
                    max_index=300000,
                    step=step,
                    batch_size=batch_size)
test_gen = generator(float_data,
                     lookback=lookback,
                     delay=delay,
                     min_index=300001,
                     max_index=None,
                     step=step,
                     batch_size=batch_size)

# This is how many steps to draw from `val_gen`
# in order to see the whole validation set:
val_steps = (300000 - 200001 - lookback) // batch_size

# This is how many steps to draw from `test_gen`
# in order to see the whole test set:
test_steps = (len(float_data) - 300001 - lookback) // batch_size

from keras.models import Sequential
from keras import layers
from keras.optimizers import RMSprop

model = Sequential()
model.add(layers.Flatten(input_shape=(lookback // step, float_data.shape[-1])))
model.add(layers.Dense(32, activation='relu'))
model.add(layers.Dense(1))

model.compile(optimizer=RMSprop(), loss='mae')
history = model.fit_generator(train_gen,
                              steps_per_epoch=500,
                              epochs=20,
                              validation_data=val_gen,
                              validation_steps=val_steps)

```
Here is what I tired to do using pytorch and skorch:

```
def generator_skorch(data, lookback, delay, min_index, max_index,
              shuffle=False, batch_size=128, step=6,steps_per_epoch = 500):
    if max_index is None:
        max_index = len(data) - delay - 1
    i = min_index + lookback
    steps = 0
    while steps < steps_per_epoch:
        if shuffle:
            rows = np.random.randint(
                min_index + lookback, max_index, size=batch_size)
        else:
            if i + batch_size >= max_index:
                i = min_index + lookback
            rows = np.arange(i, min(i + batch_size, max_index))
            i += len(rows)

        samples = np.zeros((len(rows),
                           lookback // step,
                           data.shape[-1]))
        targets = np.zeros((len(rows),))
        for j, row in enumerate(rows):
            indices = range(rows[j] - lookback, rows[j], step)
            samples[j] = data[indices]
            targets[j] = data[rows[j] + delay][1]
        yield samples, targets
        steps = steps + 1
class basic_approach(nn.Module):
    def __init__(self):
        super(basic_approach,self).__init__()
        self.model = nn.Sequential(nn.Flatten(),nn.Linear(3360,32),nn.ReLU(),nn.Linear(32, 1))
    def forward(self,X):
        return self.model(X)
  net = NeuralNetRegressor(
    basic_approach,
    max_epochs=1,
    criterion = nn.L1Loss,
    iterator_train__shuffle=True,batch_size = 128,
    optimizer = torch.optim.RMSprop,
    lr = 0.001,
    device = 'cuda',
    optimizer__alpha=0.9,
    optimizer__eps= 1e-07,warm_start = True,train_split= None,verbose=0
) 
num_epochs = 20
for epoch in range(num_epochs):
    print('Started epoch:', epoch)
    train_gen_skorch = generator_skorch(float_data,
                      lookback=lookback,
                      delay=delay,
                      min_index=0,
                      max_index=200000,
                      shuffle=True,
                      step=step, 
                      batch_size=batch_size)
    for sample,target in train_gen_skorch:
        net.fit(sample.astype(np.float32),target.astype(np.float32).reshape(-1,1))
    print(net.history[-1])

```

The code that I have is probably not most efficient way of doing it. Any idea how can I leverage pytorch dataloader to accomplish the same.

**Comment by BenjaminBossan** [comment]
I know too little about your problem to be really helpful here. Do you have memory constraints and that's why you need to use a generator or is it just the most convenient way?

In general, it's very rare that you need to implement your own `DataLoader` in PyTorch, the default one often does the job. It can also take care of concerns like sampling and shuffling, so probably you can git rid of that part of the code.

What you might want to take a look at is implementing your own custom `Dataset` class. This is the class that actually holds your data, whereas the `DataLoader` is responsible for handling of your data.

Here are some pointers:

- PyTorch [documentation](https://pytorch.org/docs/stable/data.html) about handling with data
- Example of creating a [custom dataset](https://pytorch.org/tutorials/beginner/basics/data_tutorial.html#creating-a-custom-dataset-for-your-files)

If you create your custom dataset, skorch supports fitting that. Since you return a tuple of two elements, it should even work more or less out of the box.

**Comment by sushmit86** [comment]
@BenjaminBossan  Thanks for your helpful post and apologies for not explaining my problem carefully. I think I get to have what I need. Is it possible to have two different (torch Dataset) custom datasets (one training and one validation) passed on to Skorch net? Do I need to use a predefined split?

**Comment by ottonemo** [comment]
> @BenjaminBossan Thanks for your helpful post and apologies for not explaining my problem carefully. I think I get to have what I need. Is it possible to have two different (torch Dataset) custom datasets (one training and one validation) passed on to Skorch net? Do I need to use a predefined split?

It is possible to write your own dataset and pass it to `net.fit`. In your case a [`torch.utils.data.IterableDataset`](https://pytorch.org/docs/stable/data.html?highlight=torch%20utils%20data%20iterabledataset#torch.utils.data.IterableDataset) would make sense I think. If you want to have validation data you will certainly need a predefined split / a separate validation set since it is hard to automatically split generative datasets and skorch does not support this right now.

**Comment by sushmit86** [comment]
Thanks for the help. I was able to recreate the code using Dataloader.

**Comment by BenjaminBossan** [comment]
Great. Feel free to re-open the issue if further questions come up.
